server/src/main/java/org/elasticsearch/indices/SystemIndices.java

```diff
@@ -36,34 +56,60 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 import java.util.stream.Collectors;
 
 import static java.util.stream.Collectors.toUnmodifiableList;
-import static org.elasticsearch.tasks.TaskResultsService.TASK_INDEX;
+import static org.elasticsearch.cluster.metadata.IndexMetadata.INDEX_FORMAT_SETTING;
 
 /**
  * This class holds the {@link SystemIndexDescriptor} objects that represent system indices the
  * node knows about. Methods for determining if an index should be a system index are also provided
  * to reduce the locations within the code that need to deal with {@link SystemIndexDescriptor}s.
  */
-public class SystemIndices {
+public class SystemIndices implements ClusterStateListener {
```

Comment:
Can we separate the "SystemIndexManager" out of the `SystemIndices` class? I like that `SystemIndices` is fairly lightweight and compact. The manager (and other services) will depend on `SystemIndices`, like `SystemIndexMetadataUpgradeService`.